File: Leetcode/11.盛最多水的容器.cpp
/*
 * @lc app=leetcode.cn id=11 lang=cpp
 *
 * [11] 盛最多水的容器
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-09-27-17-38*/

// from https://github.com/fengwei2002/Algorithm
/*
 !
 !
 *
 * https://leetcode-cn.com/problems/container-with-most-water/solution/lc11-fengwei2002-by-fengwei2002-kpan/
 */

// @lc code=start
class Solution {
public:
    int maxArea(vector<int>& h) {
        int ans = 0;

        int i = 0, j = h.size() - 1;
        while (i < j) {
            ans = max(ans, min(h[i], h[j]) * (j - i));
            
            if (h[i] < h[j]) i++;
            else j--;
        }
        
        return ans;
    }
};
// @lc code=end

File: Leetcode/11.盛最多水的容器.go
/*
 * @lc app=leetcode.cn id=11 lang=golang
 *
 * [11] 盛最多水的容器
 */
package leetcode
/* --- 2022-04-06-19-34 --- */
// https://github.com/fengwei2002/Algorithm

// @lc code=start
func maxArea(height []int) int {
	// 找出两条线，使得这两个线中可以放置最多的水 
	ans := 0 
	for i, j := 0, len(height) - 1; i < j; {
		ans = max(ans, min(height[i], height[j]) * (j - i))
		if height[i] > height[j] {
			j--
		} else {
			i++
		}
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	} 
	return b
}

func max(a, b int) int {
	if a < b {
		return b
	} 
	return a
}
// @lc code=end

